### PR TITLE
usability features

### PR DESCRIPTION
**Pull request by ryanjmulder**
Hello Wongoo,

 Thank you very much for the excellent sqlite3pp library. We've been using it for 3 years now, and it has worked great. Today, I went looking for a sqlite wrapper that did not require boost, and I discovered that you've already removed the boost dependency!

 As I merged your update into our code today, I noticed quite a few little things that we've tweaked in the last 3 years. I don't know if you'll want any of our tweaks, and it's ok if you don't, but in the spirit of open source software, I thought I should at least share them with you.

 If you have any questions or notice something broken, please let me know.

Thanks again,
Ryan

### CODE REVIEW

File: src/sqlite3ppext.h [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #include <cstddef>
 #include <map>
 #include <tuple>
+#include <memory>
```
Comment (iwongu) [inline src/sqlite3ppext.h]: Can you order includes alphabetically?

File: src/sqlite3ppext.cpp [hunk 2]
```diff
@@ -23,6 +23,7 @@
 // THE SOFTWARE.
 
 #include "sqlite3ppext.h"
+#include <cstring>
```
Comment (iwongu) [inline src/sqlite3ppext.cpp]: Please, put system header files before others and put a blank line between the blocks.

File: src/sqlite3pp.h [hunk 3]
```diff
@@ -282,6 +305,7 @@ namespace sqlite3pp
    private:
     database* db_;
     bool fcommit_;
+    std::unique_lock< std::mutex > lock_;
```
Comment (iwongu) [inline src/sqlite3pp.h]: When did you need the locking for a transaction?

File: src/sqlite3pp.h [hunk 4]
```diff
@@ -108,6 +123,12 @@ namespace sqlite3pp
     authorize_handler ah_;
   };
 
+  #if defined(_MSC_VER) && (_MSC_VER <= 1800)
```
Comment (iwongu) [inline src/sqlite3pp.h]: Let's not use this macro. I don't think noexcept(false) is worth to be used.

File: src/sqlite3pp.h [hunk 5]
```diff
@@ -251,6 +273,7 @@ namespace sqlite3pp
       query_iterator& operator++();
 
       value_type operator*() const;
+      value_type operator->() const;
```
Comment (iwongu) [inline src/sqlite3pp.h]: Let's not add this. The implementation seems not right.

File: src/sqlite3pp.h [hunk 6]
```diff
@@ -224,7 +246,7 @@ namespace sqlite3pp
       }
 
       getstream getter(int idx = 0);
-
+      inline rows* operator -> (){ return this; }
```
Comment (iwongu) [inline src/sqlite3pp.h]: Let's not add this.

File: src/sqlite3pp.h [hunk 7]
```diff
@@ -92,13 +101,19 @@ namespace sqlite3pp
...
+    int load( const char *filename );
+    int save( const char *filename );
+
     void set_busy_handler(busy_handler h);
     void set_commit_handler(commit_handler h);
     void set_rollback_handler(rollback_handler h);
     void set_update_handler(update_handler h);
     void set_authorize_handler(authorize_handler h);
 
+    std::mutex m_transactionMutex;
```
Comment (iwongu) [inline src/sqlite3pp.h]: When did you need this locking?

File: src/sqlite3pp.h [hunk 8]
```diff
@@ -92,13 +101,19 @@ namespace sqlite3pp
 
     int set_busy_timeout(int ms);
 
+    int load( const char *filename );
```
Comment (iwongu) [inline src/sqlite3pp.h]: Let's not add these load and save here. I prefer these done outside of the sqlite3pp class.

File: src/sqlite3pp.h [hunk 9]
```diff
@@ -174,6 +195,7 @@ namespace sqlite3pp
      private:
       command& cmd_;
       int idx_;
+      bindstream& operator = ( const bindstream& ) = delete;
```
Comment (iwongu) [inline src/sqlite3pp.h]: Please save some spaces. =)

bindstream& operator=(const bindstream&) = delete;

File: src/sqlite3pp.h [hunk 10]
```diff
@@ -27,10 +27,11 @@
 
 #include <functional>
 #include <iterator>
-#include <sqlite3.h>
+#include "sqlite3.h"
```
Comment (iwongu) [inline src/sqlite3pp.h]: Let's keep it as a system header.

File: src/sqlite3pp.h [hunk 11]
```diff
@@ -27,10 +27,11 @@
...
 #include <stdexcept>
 #include <string>
 #include <tuple>
+#include <mutex>
```
Comment (iwongu) [inline src/sqlite3pp.h]: Please sort header files.

**Comment by iwongu** [comment]
ryanjmulder, thanks for using this. I'm glad it served you well for 3 years. =)
I left some comments in your changes. I might be being a little more restrictive than you wanted. Please consider that I'm trying to make it usable while keeping it as simple as possible.
Thanks!

**Comment by iwongu** [comment]
ryanjmulder, I applied the changes you made in the master branch. I dropped the load/save methods. I might be reconsidering adding the features but it seems too much for this simple wrapper for now.
I'm closing this request. Please reopen if you have any opinions or suggestions.

Thanks again!

**Comment by ryanjmulder** [comment]
I'm sorry for my slow response, I meant to make the changes you recommended, but a work issue came up and I've been really busy. Thank you very much for all your help!

**Comment by iwongu** [comment]
No problems at all! Thanks for your suggestions!

**Pull request closed without merging**